Return upstream HTTP error code from terminal payment endpoint

anu-rock opened the pull request:
Fixes #2847

#### Changes proposed in this Pull Request

Replace `502` HTTP status code in the `capture_terminal_payment` endpoint with the one returned by Stripe.

Previously, we were returning a generic 502 HTTP status code for all Stripe API errors. While 502 was not an inappropriate choice, it represented *any* error at gateway or proxy, leading to the loss of the actual status code returned by the upstream API.

For example, when we try to charge more than the amount in payment intent Stripe fails with a 400 HTTP code which signifies that the input(s) received by Stripe were incorrect or in bad format. Ignoring 400 and returning 502
meant that our endpoint's client had to rely on the error message to determine the cause of error.

#### Testing instructions

A scenario is assumed where a user, after placing a COD order, changes their mind and wants an item removed before making payment.

1. Create a new order with two or more non-digital/virtual products. Eg. a Hoodie ( $45 ) and a Polo ( $20 ).
2. When checking out the order, select **Cash on delivery** as the payment method.
3. The order status will hopefully change to **Processing** soon after order creation. If it's On Hold, change to Processing manually.
4. Refund one item from the order, say Hoodie. The total payable amount should now be reduced by $45. ( Note: directly editing an order to remove items is not possible when the status is Processing ).
5. In WooCommerce Mobile App, attempt to collect payment for the above order. It should ultimately fail due to an incorrect payable amount ( #2846 ). Note the HTTP response from `capture_terminal_payment` endpoint in Charles Proxy (or another network traffic analyzer).
6. If testing with WooCommerce mobile app is not possible, make a direct POST request (from a REST client) to `capture_terminal_payment` endpoint with the above order's id and a new payment intent id ( the payment intent must be created* for an amount greater than the payable amount noted in step 4 ). Note the HTTP response in the REST client.
7. Before this change, the HTTP response code was 502. After this change, it should be 400. ( Screenshots below )

\* We can create a manual payment with the option Capture funds later in https://dashboard.stripe.com/test/connect/accounts/acct_YOUR_TEST_ID/payments/new

##### Before

<img width="1391" alt="image" src="https://user-images.githubusercontent.com/1288616/132537323-5a7adac8-2b92-4f3e-92c4-4098a6d2d39a.png">

##### After

<img width="1459" alt="image" src="https://user-images.githubusercontent.com/1288616/132674198-5aa820fb-c818-472a-aa70-0838a180664a.png">

-------------------

- [ ] Added changelog entry - **does not apply**
- [x] Covered with tests
- [x] Tested on mobile

anu-rock commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1664,16 +1668,18 @@ public function capture_charge( $order ) {
 				WC_Payments_Explicit_Price_Formatter::get_explicit_price( wc_price( $amount, [ 'currency' => $currency ] ), $order )
 			);
 			$order->add_order_note( $note );
+			$http_code = 500;
```
I'm not sure if `500` is the appropriate code in case of capture amount failure.

htdat replied:
I think it should be `400`, which is similar to the main issue here.

It looks like that right after the current line 1660 should have `$http_code = 400` too;

anu-rock replied:
@htdat You mean the errors on 1671 and 1660 are not server errors but client errors (invalid input data)? I do not have enough context of all possible errors possible at these locations, so I relied on the unit tests for my understanding.

Eg. in `test_capture_charge_failure` and `test_capture_charge_failure_non_usd`, the error seems to be caused by the payment status remaining the same ("requires_charge") between order creation and capturing of intent. My assumption was that the payment status would not change if the upstream Stripe API somehow failed on calling `capture_intention`.

In other tests — `test_capture_charge_api_failure`, `test_capture_charge_api_failure_non_usd`, and `test_capture_charge_expired` — a mock exception is explicitly thrown with HTTP code 500.

Btw, I missed setting `$http_code` on line 1661. Thanks for catching that!

anu-rock replied:
Actually, I now remember why I didn't set `$http_code` on line 1661. That's because it's only possible to enter that `elseif` block when `$error_message` is non-empty:

https://github.com/Automattic/woocommerce-payments/blob/f35dd3dbe6ec04e2d5fdc03a7483517ca5f70573/includes/class-wc-payment-gateway-wcpay.php#L1606-L1622

On lines 1609 and 1621, we set `$http_code` along with `$error_message`. In other words,  `$http_code` will have been set if `$error_message` is set. So, there is no need to explicitly set it again on 1661. Doing so would override the upstream API response code.

However, it was necessary to set `$http_code` on line 1671 to handle cases as described in `test_capture_charge_failure` and `test_capture_charge_failure_non_usd`.

htdat replied:
> Actually, I now remember why I didn't set `$http_code` on line 1661. That's because it's only possible to enter that `elseif` block when `$error_message` is non-empty
> [...]

Ah, that makes sense for me now. If so, nitpick: is it worth adding a comment here?

htdat replied:
**Regarding `400` or `500` status:** 

> Eg. in test_capture_charge_failure and test_capture_charge_failure_non_usd, the error seems to be caused by the payment status remaining the same ("requires_charge") between order creation and capturing of intent... 

I think this test means if an error happens when we try to capture a charge, Stripe will keep PaymentIntent's status as `requires_charge`. 

> In other tests — `test_capture_charge_api_failure`, `test_capture_charge_api_failure_non_usd`, and `test_capture_charge_expired` — a mock exception is explicitly thrown with HTTP code 500.

All $http_code places in those tests are added in your PR so I don't think it's reasonable to use `$http_code` as a judgment here.

---- 

I also looked at this PR https://github.com/Automattic/woocommerce-payments/pull/106, where this error message `A capture of %1$s <strong>failed</strong> to complete.` was added, and it's not clear when this can happen. **Therefore, I agreed that this is a grey area and I don't have a strong opinion on this anymore.** Let's wait for other reviewers' opinions on this issue.

anu-rock replied:
Thanks for your thoughts, @htdat. In that case, pinging @kalessil and @marcinbot for their thoughts on this particular thread.

htdat reviewed (commented):
Great work here. 

✅ Following test steps => looks good. 
❓ I have some comments below. 

1, I have an inline comment regarding the proper status code. 
2, Do you think it's worth creating a test for this case when Stripe response status is 400? 
3, Regarding step 6 of "Testing instructions": 

> the payment intent must be created for an amount greater than the payable amount noted in step 4. 

Actually, without Woo Mobile app, this instruction is not clear. Instead, we can create a manual payment with the option `Capture funds later` in https://dashboard.stripe.com/test/connect/accounts/acct_YOUR_TEST_ID/payments/new, which is similar to this PR when introducing `capture_terminal_payment`. 

--- 

Also, I think this PR needs at least one Approve from other reviewers as I am not too much familiar with the relevant codebase.

anu-rock commented:
> Great work here.
> 
> ✅ Following test steps => looks good.
> ❓ I have some comments below.
> 

Great points. Thanks for your review.

> 1, I have an inline comment regarding the proper status code.

I've added my observations. As per my understanding, it represents a server error rather than a client error.

> 2, Do you think it's worth creating a test for this case when Stripe response status is 400?

Done in 2feeadd ☑️ 

> 3, Regarding step 6 of "Testing instructions":
> 
> > the payment intent must be created for an amount greater than the payable amount noted in step 4.
> 
> Actually, without Woo Mobile app, this instruction is not clear. Instead, we can create a manual payment with the option `Capture funds later` in https://dashboard.stripe.com/test/connect/accounts/acct_YOUR_TEST_ID/payments/new, which is similar to this PR when introducing `capture_terminal_payment`.
> 

Updated testing instructions as per this input ☑️

htdat commented on tests/unit/admin/test-class-wc-rest-payments-orders-controller.php:
```diff
@@ -223,6 +223,52 @@ public function test_capture_terminal_payment_error_when_capturing() {
+		$this->assertInstanceOf( 'WP_Error', $response );
+		$data = $response->get_error_data();
+		$this->assertArrayHasKey( 'status', $data );
+		$this->assertEquals( 400, $data['status'] );
```
I would use` assertSame` and evaluate the error code of WP_Error instance too, something like this: 

```suggestion
		$this->assertSame( 400, $data['status'] );
		$this->assertSame( 'wcpay_capture_error', $response->get_error_code() );
```

anu-rock replied:
I had reused much of the previous unit test as a template and missed this. Thanks for the tip 🙌

htdat reviewed (commented):
I left some comments for your new test and some thoughts regarding whether `400` or `500` status for a specific case. 
If my test suggestions make sense, please update your test, then let's wait for at least another reviewer.

anu-rock commented on tests/unit/admin/test-class-wc-rest-payments-orders-controller.php:
```diff
@@ -223,6 +223,52 @@ public function test_capture_terminal_payment_error_when_capturing() {
 		$this->assertEquals( 'Payment capture failed to complete with the following message: Test error', $response->get_error_message() );
 	}
 
+	public function test_capture_terminal_payment_error_capture_amount_greater_than_authorized_amount() {
```
That makes a tonne of sense. Will update my test 👍

anu-rock commented:
@marcinbot and @kalessil Thanks for your inputs!

I'm aligned with the suggestions and have made the changes in 9923a62.
